File: codigo_base.py
################################################################################
#                                                                              #
#     Archivo: codigo_base.py                                                  #
#     Fecha de última revisión: 08/09/2023                                     #
#     Autores: Francisco Javier Pizarro 821259                                 #
#     Comms:                                                                   #
#           Este archivo contiene el esqueleto para facilitar la implementación#
#           de la solución del problema del Unicode 23                         #
#           Modificar el contenido de la función solve_problem_instance        #
#     Use:  python3 codigo_base.py                                             #
#                                                                              #
################################################################################

def solve_problem_instance(clases, Maximum):
    print("\tEjecutando el algoritmo para resolver el problema")
    nClases = len(clases)
    
    #######################################
    # IMPLEMENTAR AQUÍ VUESTRA SOLUCIÓN :D
    #######################################
    # Este bucle muestra como acceder a cada una de las cifras de alumnos de las distintas clases
    for clase in clases:  
        print(clase)

    solucion = []
    mayor_N_asistentes = 0

    return mayor_N_asistentes, solucion, len(solucion)

def read_file(nombre_fichero):
    print("\tCargando datos del fichero de entrada")
    with open(nombre_fichero, 'r') as file:
        # Lectura de los enteros de la primera linea
        first_line = file.readline().strip()
        first_ints = first_line.split()
        capacidad_maxima = int(first_ints[0])
        num_ints_to_read = int(first_ints[1])

        # Lectura de los enteros de la segunda línea separados por espacios
        second_line = file.readline().strip()
        second_ints = second_line.split()

        # Almacenamos los enteros en una lista
        clases = [int(num) for num in second_ints]

    return clases, capacidad_maxima

def write_file(nombre_fichero,num_clases_solucion, clases_solucion):
    print("\tEscribiendo resultados en el fichero de salida")
    with open(nombre_fichero, 'w') as file:
        # Escritura del número de clases que asisten a la excursión
        file.write(str(num_clases_solucion)+'\n')
        # Escritura del número de alumnos de cada clase que asisten separados por espacios
        file.write(" ".join(map(str,clases_solucion)))
    return

NOMBRES_FICHEROS = ["facil","normal","dificil","aun_mas_dificil"]

def main():
    # Ejecutamos todo para cada uno de los ficheros
    for nombre_fichero in NOMBRES_FICHEROS:
        print("Ejecutando para el dataset:",nombre_fichero)
        # Leemos los datos
        clases, capacidad_maxima = read_file("./datasets/"+nombre_fichero+ ".txt")
        # Calculamos las soluciones
        mayor_N_asistentes, clases_solucion, num_clases_solucion = solve_problem_instance(clases, capacidad_maxima)
        # Escribimos las soluciones
        write_file("./soluciones/"+nombre_fichero+ ".txt",num_clases_solucion,clases_solucion)

if __name__ == "__main__":
    main()

File: Unicode23/codigo_base.cpp
/////////////////////////////////////////////////////////////////////////////////
//                                                                             //
//    Archivo: codigo_base.cpp                                                 //
//    Fecha de última revisión: 08/09/2023                                     //
//    Autores: Francisco Javier Pizarro 821259                                 //
//    Comms:                                                                   //
//          Este archivo contiene el esqueleto para facilitar la implementación//
//          de la solución del 1º problema del Unicode 23                      //
//          Modificar el contenido de la función solve_problem_instance        //
//    Use:  g++ codigo_base.cpp                                                //
//          Lanzar el ejecutable generado                                      //
//                                                                             //
/////////////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

vector<int> solve_problem_instance(const vector<int>& clases, int Maximum) {
    cout << "\tEjecutando el algoritmo para resolver el problema" << endl;
    int nClases = clases.size();
    
    ///////////////////////////////////////
    // IMPLEMENTAR AQUÍ VUESTRA SOLUCIÓN :D
    ///////////////////////////////////////
    
    // Este bucle muestra cómo acceder a cada una de las cifras de alumnos de las distintas clases
    for (int i = 0; i < nClases; ++i) {
        cout << clases[i] << endl;
    }

    // Dato útil para los de 1º
    // para añadir algo al vector de la solución emplead una linea de código similar a  la siguiente:
    // solucion.push_back(12);
    vector<int> solucion;
    int mayor_N_asistentes = 0;

    return solucion;
}

pair<vector<int>, int> read_file(const string& nombre_fichero) {
    cout << "\tCargando datos del fichero de entrada" << endl;
    ifstream file(nombre_fichero);
    if (!file.is_open()) {
        cerr << "Error: Unable to open file " << nombre_fichero << endl;
        exit(1);
    }

    vector<int> clases;
    int capacidad_maxima;
    int num_enteros_a_leer;
    // Lectura de la capacidad máxima
    file >> capacidad_maxima;
    file >> num_enteros_a_leer;
    // Lectura de los alumnos de las distintas clases
    int num;
    while (file >> num) {
        clases.push_back(num);
    }

    return make_pair(clases, capacidad_maxima);
}

void write_file(const string& nombre_fichero, int num_clases_solucion, const vector<int>& clases_solucion) {
    cout << "\tEscribiendo resultados en el fichero de salida" << endl;
    ofstream file(nombre_fichero);
    if (!file.is_open()) {
        cerr << "Error: Unable to open file " << nombre_fichero << endl;
        exit(1);
    }

    // Escritura del número de clases que asisten a la excursión
    file << num_clases_solucion << endl;
    // Escritura del número de alumnos de cada clase que asisten separados por espacios
    for (int i = 0; i < num_clases_solucion; ++i) {
        file << clases_solucion[i];
        if (i < num_clases_solucion - 1) {
            file << " ";
        }
    }
}

int main() {
    // Ejecutamos todo para cada uno de los ficheros
    vector<string> NOMBRES_FICHEROS = {"facil", "normal", "dificil", "aun_mas_dificil"};

    for (const string& nombre_fichero : NOMBRES_FICHEROS) {
        cout << "Ejecutando para el dataset: " << nombre_fichero << endl;
        // Leemos los datos
        pair<vector<int>, int> data = read_file("./datasets/" + nombre_fichero + ".txt");
        const vector<int>& clases = data.first;
        int capacidad_maxima = data.second;

        // Calculamos las soluciones
        vector<int> clases_solucion = solve_problem_instance(clases, capacidad_maxima);
        int num_clases_solucion = clases_solucion.size();

        // Escribimos las soluciones
        write_file("./soluciones/" + nombre_fichero + ".txt", num_clases_solucion, clases_solucion);
    }

    return 0;
}